src/rgw/rgw_data_sync.cc

```diff
@@ -4061,6 +4062,34 @@ int RGWBucketFullSyncCR::operate(const DoutPrefixProvider *dpp)
     if (sync_result == 0) {
       sync_status.state = BucketSyncState::Incremental;
       tn->log(5, SSTR("set bucket state=" << sync_status.state));
+      if (sync_status.shards_done_with_gen.empty()) {
+	tn->log(5, "Generation has no shards in incremental. Updating.");
+	tn->log(5, "Fetching index log.");
+	yield call(new RGWListBucketIndexLogCR(
+		     sc, {sync_pipe.source_bucket_info.bucket, 0}, {},
+		     sync_status.incremental_gen,
+		     &extended_result));
+	if (retcode < 0) {
+	  tn->log(-1, SSTR("ERROR: Unable to fetch bilog list for zone="
+			   << sc->source_zone << " retcode="
+			   << retcode));
+	  return set_cr_error(retcode);
+	}
+	if (extended_result.next_log) {
+	  tn->log(5, "Got next_log.");
+	  sync_status.incremental_gen = extended_result.next_log->generation;
+	  sync_status.shards_done_with_gen.clear();
+	  sync_status.shards_done_with_gen.resize(
+	    extended_result.next_log->num_shards);
+	  tn->log(5, SSTR("Updated gen to " << sync_status.incremental_gen
+			  << " and num_shards to "
+			  << sync_status.shards_done_with_gen.size()));
+	} else {
+	  tn->log(0, "WARNING: Current generation does not exist, but "
+		  "unable to find new one.");
+	}
+      }
+
```

Comment:
@cbodley Please check this part as it's a change to the main sync logic.

Reply:
`InitBucketFullSyncStatusCR` gives an empty `shards_done_with_sync` if it's provided an empty `rgw_bucket_index_marker_info`. It also sets the sync status to Full rather than Init.

Fetching the `rgw_bucket_index_marker_info` for `bucket sync init` surprisingly doesn't work, due to [a condition in RGWSyncBucketCR](https://github.com/ceph/ceph/blob/d47699bda391474f5e8e742b50e9d6e7571912f2/src/rgw/rgw_data_sync.cc#L5191-L5198) that only reads the `rgw_bucket_index_marker_info` in states of `Init` and `Stopped`, so that the call to `InitBucketFullSyncStatusCR` gets called with an empty one in the case of Full sync.

I don't think this is correct behavior as it means the code where we write data and the code where we get the data to write are gated differently.

However, changing `bucket sync init` to write an empty status with state `Init` allows `bucket sync run` to complete.

Reply:
Made a patch to address this fault in commit a640583fcba974eff4b24735efc1bc12b046d676 . It's not necessary to pass the tests, but I think the behavior is more robust and correct.